Added BaseService and OrderCreationService

abhishekgupta5 opened the pull request:
**Brief:** https://linear.app/nebulab-retainers/issue/ABU-6/create-paytomorrow-order

- [x] Added code to create payTomorrow order from console with valid access token
- [x] Specs

**Usage ->**
Prerequisite -
1. Make sure you have a **SolidusPayTomorrow::PaymentMethod** with valid username, password and signature. This can be done from the UI via the sandbox app or rails console
2. Make sure you have an order with **total > 500** 

Once yo have these 2, do this in rails console to create an order on PayTomorrow

```ruby
# Given that you have `order` and `payment_method` vars set
SolidusPayTomorrow::Client::CreateOrderService.call(order:, payment_method:)
```

This should return with the checkout URL and order token

abhishekgupta5 committed:
Added generic dependencies for codebase

* HTTParty for making http requests
* pry-byebug for debugging

DanielePalombo commented on app/services/solidus_pay_tomorrow/client/create_order_service.rb:
```diff
@@ -0,0 +1,62 @@
+
+      def create_body
+        { orderId: order.number,
+          # TODO: We'll need to implement firstname and lastname. They are deprecated in Solidus
```
You can use this [model](https://github.com/solidusio/solidus/blob/master/core/app/models/spree/address/name.rb) to retrieve firstname and lastname

abhishekgupta5 replied:
Used in https://github.com/nebulab/solidus_pay_tomorrow/pull/7/commits/390014a30a052261cae08adcd7221d7641aaf79f

DanielePalombo commented on app/services/solidus_pay_tomorrow/client/create_order_service.rb:
```diff
@@ -0,0 +1,62 @@
+          applicationItems: items }
+      end
+
+      # TODO: Revisit and see if we need any more attributes
```
These look enough to me

DanielePalombo commented on app/services/solidus_pay_tomorrow/client/create_order_service.rb:
```diff
@@ -0,0 +1,62 @@
+          state: order.bill_address.state.abbr,
+          email: order.email,
+          # TODO: Update URLs with actual routes once controllers are written
+          returnUrl: 'https://success-redirect-url.com',
+          cancelUrl: 'https://cancelled-order-redirect-url.com',
+          notifyUrl: 'https://webhook-notify-url.com',
```
Probably the domain will be the same every time. We can consider using just a constant for the path

abhishekgupta5 replied:
Used in https://github.com/nebulab/solidus_pay_tomorrow/pull/7/commits/390014a30a052261cae08adcd7221d7641aaf79f

DanielePalombo commented:
Can we prioritize this PR? I would like to use it to create the application, so that we can test the other gateway services(capture/void etc..) on the sandbox.

abhishekgupta5 committed:
Added SolidusPayTomorrow::Client::BaseService

All other client related services will inherit from
this BaseService

abhishekgupta5 committed:
Added OrderCreationService

* Takes Spree::Order as argument, creates a payTomorrow order
  using the API -> https://docs.paytomorrow.com/docs/api-reference/api/create-order/
  and returns an order token along with a URL

* Client specs

DanielePalombo commented on spec/services/solidus_pay_tomorrow/client/create_order_service_spec.rb:
```diff
@@ -0,0 +1,47 @@
+      expect(response).to eq(success_response)
+    end
+
+    def success_response
```
What is the benefit to having it as a method, rather than making it explicit in the example with:

```
expect(response).to eq(
  url: "https://consumer-staging.paytomorrow.com/verify/personal?app=11171280-8e2a-4b2e-8855-76285fc578c6&auth=e3de6c12-ed9a-4bb1-a066-5c62a1ba19d4",
  token: "11171280-8e2a-4b2e-8855-76285fc578c6"
)
```

abhishekgupta5 replied:
**Clarification:**
This is also being used on **Line:8** in the same file.
Hence it makes sense to have it at one place and use it at both these places.

**Variable vs method**
Since this is being used at 2 different places with different scopes, it can only live in a top level let, an instance variable or a method.

1. A let is still a method. And since this is static info, this should not be in let (no need for memoization)
2. Having instance variable is discouraged anyway in specs. Moreover, rubocop complains with - "prefer a local variable, let or a method rather than instance variable"

Hence, method makes sense here. That's why it is here.
If this is not fine, let me know what do we want here

abhishekgupta5 replied:
Fixed as per discussion on call in https://github.com/nebulab/solidus_pay_tomorrow/pull/7/commits/75dc5280c554494a2fd5784ac6c45656b06e9fdc

DanielePalombo commented on app/services/solidus_pay_tomorrow/client/create_order_service.rb:
```diff
@@ -0,0 +1,66 @@
+
+      CREATE_ENDPOINT = 'api/application/ecommerce/orders'
+      # TODO: Update URI once implemented
+      WEBHOOK_URL = 'https://domain.com/webhooks/:type'
```
Considering the URL will depend on the domain, probably it shouldn't be a constant, but a method.

abhishekgupta5 replied:
**Confirmation:** Will this be fine?

```ruby
def webhook_url(type)
  "https://domain.com/webhooks/#{type}"
end
```

abhishekgupta5 replied:
Fixed in https://github.com/nebulab/solidus_pay_tomorrow/pull/7/commits/75dc5280c554494a2fd5784ac6c45656b06e9fdc

DanielePalombo commented on spec/services/solidus_pay_tomorrow/client/create_order_service_spec.rb:
```diff
@@ -0,0 +1,47 @@
+
+  before do
+    url = 'https://api-staging.paytomorrow.com/api/application/ecommerce/orders'
+    headers = { 'Authorization': "Bearer #{ENV['ACCESS_TOKEN']}",
```
What is this? 

AFAIK `ACCESS_TOKEN` can't be stored on the env variable but we should allow to the client to refresh it anytime we need it.

abhishekgupta5 replied:
The refresh functionality is out of the scope of this PR. On Friday's call, we decided to do it in another PR when needed.

**Question:** If not env variable, where should ACCESS_TOKEN be? At this point of time.

abhishekgupta5 replied:
Added the functionality to always get valid token in this fixup [commit](https://github.com/nebulab/solidus_pay_tomorrow/pull/7/commits/b4915dbdae56e8587c7fc5a2621aa3f945fd3138)

abhishekgupta5 replied:
One problem I see with this is - all the services' specs that'll inherit from BaseService will have to also stub the access token call. Like - **OrderCreationService**

Since access_token functionality need not be tested in **OrderCreationServiceSpec** , to avoid unnecessary stubbing I have stubbed with _allow_any_instance_of_ which goes against rubocop but keeps the access_token stubbing to a minimal.
Wdyt? @DanielePalombo . Any better way in mind?

DanielePalombo replied:
I can't think of any better way at the moment.
But we have to test that feature somewhere.

abhishekgupta5 replied:
Yeah, added `base_service_spec.rb` [here](https://github.com/nebulab/solidus_pay_tomorrow/pull/7/commits/18cfa4e447f7d1a441d487d2dd251744ec7af351#diff-5f279aa0ae27af2af4362dcad4cb67195da4780560e606b1d018f13e729a6c0d)

abhishekgupta5 committed:
Added api_signature to payTomorrow credentials

api_signature is the basic auth token that is used to get
access_token. api_signature is present on PT merchant's
dashboard along with username and password.
Admin should enter it in Solidus Backend along with username and
password
Ref: https://docs.paytomorrow.com/docs/api-reference/api/authenticate

DanielePalombo commented on app/services/solidus_pay_tomorrow/client/base_service.rb:
```diff
@@ -0,0 +1,81 @@
+
+      # Tokens are valid for 10-12 hours, if they are near expiry, get new token
+      def current_token_invalid?
+        self.class.current_token.nil? || Time.zone.now > self.class.current_token[:expires_at] - 10.minutes
```
Why this refreshes the token 10 minutes before the expiration? 
I think we just need to refresh it when it is expired.

abhishekgupta5 replied:
Hmmm, I am just thinking of a race condition where this(_Time.zone.now > self.class.current_token[:expires_at]_) condition returns false during code execution, it becomes true just after and we do the request with expired token.
**Example** - current token is going to be expired 1 second after executing this code.

Hence, I think getting a new token with a margin would be better, wdyt?

Although, this time gap would be in < a second maybe so I think instead of 10 minutes, 1 minute or few seconds should be better?

DanielePalombo replied:
It makes sense. We can leave 10 minutes for it. Thank you for the explaination

DanielePalombo commented on app/services/solidus_pay_tomorrow/client/base_service.rb:
```diff
@@ -0,0 +1,81 @@
+        raise StandardError, result.parsed_response['error']
+      end
+
+      def valid_token
+        new_token if current_token_invalid?
+        self.class.current_token[:access_token]
+      end
+
+      def new_token
+        result = handle_errors!(HTTParty.post(token_uri, headers: token_headers, body: URI.encode_www_form(token_body)))
+        self.class.current_token =
+          { expires_at: Time.zone.now + result['expires_in'].seconds }.merge!(result.symbolize_keys)
+      end
+
+      # Tokens are valid for 10-12 hours, if they are near expiry, get new token
+      def current_token_invalid?
+        self.class.current_token.nil? || Time.zone.now > self.class.current_token[:expires_at] - 10.minutes
+      end
+
+      def token_uri
+        "#{api_base_url}/#{AUTH_ENDPOINT}"
+      end
+
+      def token_headers
+        { 'Accept': 'application/json',
+          'Content-Type': 'application/x-www-form-urlencoded',
+          'Authorization': "Basic #{payment_method.preferences[:signature]}" }
+      end
+
+      def token_body
+        { grant_type: 'password',
+          scope: 'openid',
+          username: payment_method.preferences[:username],
+          password: payment_method.preferences[:password] }
+      end
+    end
```
Where is the test of this part?

abhishekgupta5 replied:
Yes, I'll push a commit for it in a bit.

abhishekgupta5 replied:
Added [here](https://github.com/nebulab/solidus_pay_tomorrow/pull/7/commits/18cfa4e447f7d1a441d487d2dd251744ec7af351#diff-5f279aa0ae27af2af4362dcad4cb67195da4780560e606b1d018f13e729a6c0d)

DanielePalombo commented on app/services/solidus_pay_tomorrow/client/base_service.rb:
```diff
@@ -0,0 +1,81 @@
+      end
+
+      class << self
+        attr_accessor :current_token
```
I don't think the current_token should be a public method.
